src/Microsoft.VisualStudio.ProjectSystem.Managed/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshot.cs

```diff
@@ -174,35 +174,160 @@ void Add(AddDependencyContext context, IDependencyModel dependencyModel)
+                _hasReachableVisibleUnresolvedById = null;
+            }
+
+            return;
 
-                if (!dependency.Resolved && dependency.Visible)
+            static (ImmutableArray<IDependency> TopLevelDependencies, Dictionary<string, IDependency> topLevelDependencyByPath, bool hasVisibleUnresolvedDependency) Scan(ImmutableDictionary<string, IDependency> dependenciesWorld, ITargetFramework targetFramework)
+            {
+                // TODO use ToImmutableAndFree?
+                ImmutableArray<IDependency>.Builder topLevelDependencies = ImmutableArray.CreateBuilder<IDependency>();
+
+                bool hasVisibleUnresolvedDependency = false;
+                var topLevelDependencyByPath = new Dictionary<string, IDependency>(StringComparer.OrdinalIgnoreCase);
+
+                foreach ((string id, IDependency dependency) in dependenciesWorld)
                 {
-                    hasVisibleUnresolvedDependency = true;
+                    System.Diagnostics.Debug.Assert(
+                        string.Equals(id, dependency.Id),
+                        "dependenciesWorld dictionary entry keys must match their value's ids.");
+
+                    if (!dependency.Resolved && dependency.Visible)
+                    {
+                        hasVisibleUnresolvedDependency = true;
+                    }
+
+                    if (dependency.TopLevel)
+                    {
+                        topLevelDependencies.Add(dependency);
+
+                        if (!string.IsNullOrEmpty(dependency.Path))
+                        {
+                            topLevelDependencyByPath.Add(
+                                Dependency.GetID(targetFramework, dependency.ProviderType, dependency.Path),
+                                dependency);
+                        }
+                    }
                 }
 
-                if (dependency.TopLevel)
+                return (topLevelDependencies.ToImmutable(), topLevelDependencyByPath, hasVisibleUnresolvedDependency);
+            }
+
+            IReadOnlyDictionary<string, bool>? FindReachableVisibleUnresolvedDependencies()
+            {
+                // It is possible that there exists a dependency which is both visible and unresolved, yet not
+                // actually present in the tree because one of its ancestors is not visible. Therefore instead
+                // of scanning all dependencies in the snapshot, we walk the dependency graph starting with
+                // top-level visible nodes, and remember those which have at least one visible, unresolved and
+                // reachable descendant.
+
+                bool hasReachableVisibleUnresolvedDependency = false;
+
+                var hasReachableVisibleUnresolvedById = new Dictionary<string, bool>(StringComparer.OrdinalIgnoreCase);
+
+                // 'spine' is a stack containing an enumerator for each level of the graph, which is updated as
```

Comment:
What makes you say its a stack?

Reply:
It behaves as a stack. I couldn't use `Stack<T>` without allocating enumerators on the heap. Using the array allows struct enumerators to be updated via `ref` locals.